### PR TITLE
fix(vectors): reject L2 vectors whose sum-of-squares overflows f32 (BUG-386)

### LINKED ISSUE #386: fix(vectors): L2 vectors accept components whose squared magnitudes sum overflows f32, silently dropping docs from results
## Summary

`collect_vector_value` (`searchlite-core/src/index/segment.rs:527-542`) guards **cosine**-indexed vectors against sum-of-squares overflow (BUG-384), but applies no symmetric guard to **L2**-indexed vectors. The read-side query validation in `plan_vectors` (`searchlite-core/src/api/reader.rs:504-527`) has the same asymmetry: the `sum_sq.is_finite()` check only fires when `schema_field.metric` is `Cosine`.

Each vector component is individually finite (passes BUG-330 / BUG-340), but `(v_i)^2` summed across the dimensions can still overflow `f32::MAX` (e.g. `[3e19, 3e19]` — `(3e19)^2 + (3e19)^2 = 1.8e39 > f32::MAX ≈ 3.4e38`). For L2 that overflow never surfaces at write or query time; it only materialises at distance-computation time, when `l2_distance` accumulates `(x - y)^2` and the sum saturates to `+inf`. `-l2_distance` then returns `-inf`, and the doc is silently dropped further down the pipeline or ranks last inside HNSW.

Same class of bug as BUG-384 (cosine), just on the path that was not covered by that fix.

## Affected code

- `searchlite-core/src/index/segment.rs:527-542` — `collect_vector_value` gates the sum-of-squares check on `matches!(vf.metric, VectorMetric::Cosine)`; L2 falls through with no analogous check.
- `searchlite-core/src/api/reader.rs:504-527` — `plan_vectors` gates the query-side sum-of-squares check on the same `Cosine` match; L2 query vectors with overflowing pairwise magnitudes reach `metric_similarity` unvalidated.
- `searchlite-core/src/vectors/mod.rs:104-111` — `l2_distance` accumulates `d * d` without any overflow guard and returns `+inf` unchanged; `metric_similarity(L2, ...)` negates it to `-inf`.

## Reproduction / evidence

Schema with an L2 vector field, `dim: 2`:

```json
{ "embedding": { "type": "array", "items": { "type": "number" },
                 "searchlite:vector": { "dim": 2, "metric": "L2" } } }
```

**Write-side trigger (silent corruption of the persisted vector):**

1. Index a doc with `{"embedding": [3.0e19, 3.0e19]}`. Each component is finite (`3e19 < f32::MAX ≈ 3.4e38`) and passes the BUG-330 per-component guard, so `collect_vector_value` accepts the vector and the segment captures it verbatim.
2. Query with `{"vector": [0.0, 0.0], "k": 10, ...}` — components finite, passes BUG-340.
3. `l2_distance([3e19, 3e19], [0, 0])` computes `(3e19)^2 + (3e19)^2 = 1.8e39` which overflows to `+inf`; `sqrt(+inf) = +inf`; `metric_similarity(L2) = -inf`.
4. HNSW returns `(doc_id, -inf)`; `vscore *= boost` keeps it at `-inf`; the candidate is inserted into `vector_scores[(seg, doc)] = -inf`.
5. `compute_hybrid_score` guard at `reader.rs:231` (BUG-328) catches the non-finite `blended_sum` / `vector_sum` and returns `None`, so the caller simply never sees the doc — no error, no hit, no diagnostic.

**Query-side trigger (silent failure at search time only):**

1. Index a normal doc: `{"embedding": [0.0, 0.0]}`.
2. Search with `{"vector": [3.0e19, 3.0e19], "metric: L2"}`. Each query component is finite, passes BUG-340.
3. Same overflow chain produces `metric_similarity = -inf`; the doc is silently dropped from the top-k instead of being returned with a large finite distance (or an actionable error).

This is the same failure mode BUG-384 fixed for cosine: each input is individually within range, but the pairwise arithmetic combination overflows and poisons downstream scoring / neighbor selection in a way the caller cannot detect.

## Expected behavior

Symmetric with BUG-384: reject any L2-indexed or L2-queried vector whose sum-of-squares overflows `f32` at the validation boundary, with an actionable error that tells the caller to reduce component magnitudes — instead of silently dropping the doc or quietly corrupting HNSW neighbor selection.

For L2 the sum-of-squares of a single vector `v` bounds `l2_distance(v, 0)`; if that overflows, the vector is already unreachable from at least one valid query (the origin). Rejecting it at the boundary is therefore no more restrictive than the current silent-drop behaviour, and it gives the caller a diagnosable failure mode.

## Actual behavior

- Write path: the doc is persisted with finite components that nevertheless produce `+inf` distances against a range of otherwise-valid queries. HNSW graph construction sorts overflowing pairs with `-inf` at the bottom, distorting neighbor selection for other docs indexed after the offending one.
- Query path: an L2 query vector with overflowing sum-of-squares silently returns an empty / wrong hit set because every computed distance is `+inf`, every similarity is `-inf`, and the BUG-328 hybrid guard drops every candidate.

## Root cause

`collect_vector_value` and `plan_vectors` both gate the sum-of-squares overflow check on `matches!(vf.metric, VectorMetric::Cosine)` because that check was introduced to guard `normalize_in_place`. For L2 the overflow is exactly as user-visible and exactly as unactionable — it simply manifests inside `l2_distance` instead of inside `normalize_in_place` — but the symmetric guard was never added.

## Suggested fix

Move the sum-of-squares check out of the cosine-only branch in both sites, apply it to L2 as well, and use a metric-neutral error message. Add a lightweight overflow guard inside `l2_distance` itself as belt-and-braces for any code path that bypasses validation (explanations, HNSW construction-time similarity, future callers).

```rust
// searchlite-core/src/index/segment.rs — collect_vector_value
let sum_sq = vecvals.iter().map(|v| v * v).sum::<f32>();
if !sum_sq.is_finite() {
  bail!(
    "vector field {field} has components whose squared magnitudes overflow f32; reduce component magnitudes"
  );
}
if matches!(vf.metric, VectorMetric::Cosine) {
  normalize_in_place(&mut vecvals);
}
```

```rust
// searchlite-core/src/api/reader.rs — plan_vectors
let sum_sq = query_vec.iter().map(|v| v * v).sum::<f32>();
if !sum_sq.is_finite() {
  bail!(
    "vector query for field `{}` has components whose squared magnitudes overflow f32",
    vector_query.field
  );
}
if matches!(schema_field.metric, crate::index::manifest::VectorMetric::Cosine) {
  normalize_in_place(&mut query_vec);
}
```

## Environment

- Commit: 7e69ddf (origin/main)
- Rust MSRV in repo: 1.88.0
- Found by: scheduled bug hunt, follow-up to BUG-330 / BUG-340 / BUG-384 on the uncovered L2 side.

### PR DESCRIPTION
**Pull request by davidkelley**
## Summary

Fixes BUG-386 (#386). BUG-384 added a sum-of-squares overflow guard for **cosine**-indexed and cosine-queried vectors but never extended the same check to the **L2** path. Every component is individually finite (passes the BUG-330 write-side and BUG-340 query-side per-component guards), yet their squared magnitudes can still sum past `f32::MAX` — e.g. `[3e19, 3e19]` gives `1.8e39`. For L2 the overflow never surfaces at write or query time; it only materialises inside `l2_distance`, which accumulates `(x - y)^2` into `+inf`, so `-l2_distance` returns `-inf`. The BUG-328 guard in `compute_hybrid_score` then silently drops the candidate — the caller sees an empty hit set for a doc they just indexed with no actionable error.

Same class of bug as BUG-384; just on the path that fix didn't cover.

## Changes

- **`searchlite-core/src/index/segment.rs`** (`collect_vector_value`)
  - Move the sum-of-squares `is_finite()` check out of the `matches!(vf.metric, VectorMetric::Cosine)` branch so it runs for every metric, and drop the cosine-only phrasing from the error message. Normalization is still cosine-only; only the validation guard is unified.
- **`searchlite-core/src/api/reader.rs`** (`plan_vectors`)
  - Same structural change on the query side: the sum-of-squares guard now runs for every metric before the metric-specific normalization, with a metric-neutral error message (`"vector query for field ... has components whose squared magnitudes overflow f32"`).

### Tests (`searchlite-core/tests/vector_search.rs`)

- New regression `commit_rejects_l2_vector_with_overflowing_sum_of_squares` — a 3.0e19 / 3.0e19 L2 doc must fail at commit time instead of being persisted as a silently-unsearchable vector.
- New regression `l2_query_vector_with_overflowing_sum_of_squares_is_rejected` — an L2 query vector of the same shape must surface a plan-time error instead of silently returning zero hits. Stashed the src fix locally and confirmed this test's output before the fix is `Ok(SearchResult { hits: [] })` — the exact silent-empty-results symptom described in #386.
- New boundary `l2_vector_with_finite_sum_of_squares_round_trips` — `[1e19, 0.0]` (sum-of-squares `1e38 < f32::MAX`) must still round-trip, so the new guard cannot over-reject.
- Renamed and retuned `commit_accepts_vector_component_at_f32_max` → `commit_accepts_vector_component_at_f32_max_representable_sum_of_squares`, using `[1e19, 0.0]` instead of `[f32::MAX, 0.0]` since `(f32::MAX)^2` itself overflows — the "legitimate boundary" the test was defending is now the sum-of-squares boundary, not the per-component boundary.
- Retuned `vector_query_with_finite_boundary_components_is_accepted` to `[1e19, 0.0]` for the same reason.
- Updated the BUG-384 cosine regressions (`commit_rejects_cosine_vector_with_overflowing_sum_of_squares`, `cosine_query_vector_with_overflowing_sum_of_squares_is_rejected`) to assert the unified `"squared magnitudes overflow"` error shape.

## Test plan

- [x] `cargo test -p searchlite-core --features vectors --test vector_search` — all 23 vector integration tests pass, including the 3 new BUG-386 regressions.
- [x] `cargo test -p searchlite-core --features vectors` — full core matrix green (520 lib tests + every integration suite).
- [x] `cargo clippy -p searchlite-core --features vectors --all-targets -- -D warnings` — clean.
- [x] `cargo fmt --all --check` — clean.
- [x] Reverted the src changes locally and confirmed both new L2 regressions fail (the query-side test explicitly returns `Ok(hits: [])` pre-fix, proving the silent-drop symptom) while the existing cosine / per-component guards still pass — the tests genuinely exercise the bug rather than tautologically passing.

Closes #386

### CODE REVIEW

**Comment by davidkelley** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: searchlite-core/tests/vector_search.rs [hunk 1]
```diff
@@ -751,9 +751,13 @@ fn commit_rejects_vector_component_overflowing_f32_to_negative_inf() {
 }
 
 #[test]
-fn commit_accepts_vector_component_at_f32_max() {
-  // The boundary case: a value that fits in f32 (exactly f32::MAX cast to
-  // f64) must still be accepted to avoid over-rejecting legitimate inputs.
+fn commit_accepts_vector_component_at_f32_max_representable_sum_of_squares() {
```
[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/tests/vector_search.rs]: Good catch — renamed to `commit_accepts_vector_component_with_finite_squared_magnitude` in 2ac1a76 so the name tracks what's actually being asserted (the vector is `[1e19, 0.0]`, not `f32::MAX`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: searchlite-core/tests/vector_search.rs [hunk 2]
```diff
@@ -751,9 +751,13 @@ fn commit_rejects_vector_component_overflowing_f32_to_negative_inf() {
...
+  // The boundary case for BUG-330: an individually-finite component that also
+  // keeps the BUG-386 sum-of-squares inside `f32::MAX` must still be accepted
+  // to avoid over-rejecting legitimate inputs. `(1e19)^2 = 1e38 < f32::MAX ≈
+  // 3.4e38`, so this vector passes both guards. A vector with `f32::MAX`
+  // itself as a component is covered by the dedicated BUG-386 overflow tests
+  // below, since `(f32::MAX)^2` overflows.
```
[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/tests/vector_search.rs]: Agreed — the old wording was stale. Rewrote the comment in 2ac1a76 to explicitly state that the per-component `f32::MAX` boundary is no longer the effective acceptance bound post-BUG-386 (since `(f32::MAX)^2` overflows), and to note that no dedicated test exercises a literal `f32::MAX` component because any such vector is now correctly rejected by the new guard at both commit and query time. The comment now matches what the `*_overflowing_sum_of_squares_*` tests actually assert.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: searchlite-core/tests/vector_search.rs [hunk 3]
```diff
@@ -917,14 +918,16 @@ fn vector_query_with_finite_boundary_components_is_accepted() {
...
-  // f32::MAX and f32::MIN are finite boundary values; the validation guard
-  // must not reject them. The downstream L2 distance may saturate to INF
-  // inside `metric_similarity`, but that's a separate concern — this test
-  // only asserts that the request passes the finitude guard.
+  // Use a large-but-finite component that keeps the sum-of-squares within
+  // `f32::MAX` (post-BUG-386): `(1e19)^2 = 1e38 < f32::MAX ≈ 3.4e38`. The
+  // BUG-340 per-component guard allows the full f32 range, and the BUG-386
+  // sum-of-squares guard allows any vector whose squared magnitudes fit in
+  // `f32::MAX`. Values where `sum-of-squares` itself overflows are exercised
+  // by the dedicated `*_overflowing_sum_of_squares_*` tests below.
```
[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/tests/vector_search.rs]: Correct — the old comment overclaimed. Tightened in 2ac1a76 to describe the sum-of-squares acceptance boundary the test now exercises (`[1e19, 0.0]` → `(1e19)^2 = 1e38 < f32::MAX`) and to cross-reference the `*_overflowing_sum_of_squares_*` tests for the overflow side of the boundary. Dropped the claim about the per-component guard allowing the full f32 range, since that's no longer the effective bound post-BUG-386.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: searchlite-core/src/api/reader.rs [hunk 4]
```diff
@@ -502,27 +502,30 @@ impl IndexReader {
...
+      // garbage / zeroed query vectors. For L2 the same overflow manifests a
+      // step later: `l2_distance(q, doc)` accumulates `(q_i - doc_i)^2` into
+      // `+inf`, `metric_similarity(L2) = -inf`, and `compute_hybrid_score`
+      // drops every candidate via the BUG-328 guard — the caller silently
+      // sees an empty hit set with no actionable diagnostic. Guarding both
+      // metrics uniformly turns that into a plain validation error.
+      let sum_sq = query_vec.iter().map(|v| v * v).sum::<f32>();
+      if !sum_sq.is_finite() {
+        bail!(
+          "vector query for field `{}` has components whose squared magnitudes overflow f32",
```
[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/src/api/reader.rs]: Good catch — the two messages were inconsistent. Unified both write- and query-side errors in 2ac1a76 under `"... has components whose sum-of-squares overflows f32; reduce component magnitudes"`, matching BUG-384's original wording and giving query-side callers the same actionable remediation hint. Test assertions updated to match (`err.contains("sum-of-squares overflows")`).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by davidkelley** [comment]
@codex review

---
_Generated by [Claude Code](https://claude.ai/code)_
